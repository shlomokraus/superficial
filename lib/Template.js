"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var path_1 = __importDefault(require("path"));
var fs_1 = __importDefault(require("fs"));
var Templates;
(function (Templates) {
    Templates["comment"] = "comment.md";
    Templates["updated"] = "updated.md";
    Templates["revert"] = "revert.md";
    Templates["fileItem"] = "file-item.md";
    Templates["errors"] = "errors.md";
})(Templates = exports.Templates || (exports.Templates = {}));
var Template = /** @class */ (function () {
    function Template() {
    }
    Template.get = function (template) {
        var file = path_1.default.resolve("./src/templates/" + template);
        var str = fs_1.default.readFileSync(file, "utf8");
        return str;
    };
    return Template;
}());
exports.Template = Template;
