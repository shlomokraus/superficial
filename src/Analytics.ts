import UA from "universal-analytics";

export class Analytics {
  private readonly ua;

  constructor(id?: string) {
    this.ua = new UA(id);
  }

  event(...args) {
    return this.ua.event.apply(this.ua, args).send();
  }

  exception(...args) {
    return this.ua.exception.apply(this.ua, args).send();
  }

  set(...args) {
    return this.ua.set.apply(this.ua, args);
  }

  timing(...args) {
    return this.ua.timing.apply(this.ua, args).send();
  }
}
