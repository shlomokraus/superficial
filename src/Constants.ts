export const BOT_IDENTIFIER = "superficial-bot[bot]";
export const VALID_EXTENSIONS = ["ts", "js", "tsx", "jsx", "json"];
export const SCRIPT_EXTENSIONS = ["ts", "js", "tsx", "jsx"];

export enum CommentTags {
  Main = "<!-- superficial-bot-main-comment -->",
  Update = "<!-- superficial-bot-update-comment -->"
}
