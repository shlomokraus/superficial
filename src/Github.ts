import { Context } from "probot";
import { PullRequestsGetResponse } from "@octokit/rest";

export class GithubHelper
 {
  private readonly context: Context;
  private readonly pr: PullRequestsGetResponse;

  constructor(context: Context, pr: PullRequestsGetResponse) {
    this.context = context;
    this.pr = pr;
  }
  
  getFileContent = async (path: string, ref: string) => {
        const result = await this.getFile(path, ref);
        const content = Buffer.from(result.content, "base64").toString();
        return content;
      };
  

  getFile = async (path: string, ref: string) => {
    const repo = this.context.repo({ path, ref });
    const result = await this.context.github.repos.getContent(repo);
    return result.data;
  };
  
  getFiles = async () => {
    const filesRaw = await this.context.github.pullRequests.getFiles(
      this.context.repo({ number: this.pr.number})
    );
    const files = filesRaw.data.map(file => file.filename);
    return files;
  };

  getRef = async () => {
      const ref = await this.context.github.gitdata.getReference(this.context.repo({ref:"heads/"+this.pr.head.ref}));
      return ref.data;
  }

  getTree = async () => {
      const ref = await this.getRef();
      const tree = await this.context.github.gitdata.getTree(
        this.context.repo({ tree_sha: ref.object.sha })
      );
      return tree.data;
  }

  createBlob = async (content: string) => {
    const blob = await this.context.github.gitdata.createBlob(
        this.context.repo({
            number: this.pr.number,
        content,
        encoding: 'base64'
      }));
      return blob.data;
  };

  createCommit = async (files: { path: string, content: string}[]) => {

    const context = this.context;
    const github = context.github;

    const basetree = await this.getTree();
    const blobDict = {} as any;

    await Promise.all(
        files.map(async file => {
        const blob = await this.createBlob(file.content);
          blobDict[file.path] = blob.sha;
          return
    }));
    
    const tree = files.map(function(file, index) {
        return {
          path: file.path,
          mode: "100644",
          type: 'blob',
          sha: blobDict[file.path]
        };
      });

    const newTree = await github.gitdata.createTree(context.repo({
        base_tree: basetree.sha,
        tree: tree as any
    }));
      const commit = await github.gitdata.createCommit(context.repo({message: "Removing files that only have style changes", tree: newTree.data.sha, parents: [basetree.sha]}));
      await github.gitdata.updateReference(context.repo({number: this.pr.number, sha: commit.data.sha, ref: "heads/"+this.pr.head.ref }));
  };
  
}
