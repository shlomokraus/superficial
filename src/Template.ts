import path from "path";
import fs from "fs"

export enum Templates {
    comment = "comment.md",
    updated = "updated.md",
    revert = "revert.md",
    fileItem = "file-item.md",
    errors = "errors.md"
}

export class Template {
  
  static get(template: Templates) {
    const file = path.resolve(`./src/templates/${template}`)
    const str = fs.readFileSync(file,  "utf8");
    return str;
  }
}
