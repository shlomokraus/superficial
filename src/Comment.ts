import { Template, Templates } from "./Template";
import { template } from "lodash";
import { CommentTags, BOT_IDENTIFIER } from "./Constants";
import { GithubHelper } from "./Github";

// const FILE_ITEM_REGEX = /- \[(.)\] (.*)<--file-->\|/g;

export class CommentHelper {
  private readonly githubHelper: GithubHelper;

  constructor(githubHelper) {
    this.githubHelper = githubHelper;
  }

  postComment = async (
    items: { file: string; valid: boolean }[],
    errors: { file: string; error: string }[] = []
  ) => {
    const files = items.map(item => item.file);
    let comment = await this.compileComment(files, errors);

    const existing = await this.getBotMainComment();
    if (existing) {
      if (files.length === 0) {
        comment = this.tagComment(
          Template.get(Templates.updated),
          CommentTags.Main
        );
      }
      await this.githubHelper.editComment(comment, String(existing.id));
    } else {
      if (files.length > 0) {
        await this.githubHelper.createComment(comment);
      }
    }
  };

  postRevertComment = async (files: string[]) => {
    const fileItemsTemplate = Template.get(Templates.fileItem);
    const renderedFiles = files.map(filename =>
      template(fileItemsTemplate)({ filename })
    );
    const commentTemplate = Template.get(Templates.revert);
    let comment = template(commentTemplate)({
      files: renderedFiles.join("\n")
    });

    comment = this.tagComment(comment, CommentTags.Update);
    await this.githubHelper.createComment(comment);
  };

  getBotMainComment = async () => {
    const comments = await this.githubHelper.getComments();
    const filtered = comments.find(comment => {
      const isBot = comment.user.login === BOT_IDENTIFIER;
      if (!isBot) {
        return false;
      }

      const body = comment.body;
      return body.indexOf(CommentTags.Main) >= 0;
    });
    return filtered;
  };

  async compileComment(files: string[], errors: any[]) {
    const fileItemsTemplate = Template.get(Templates.fileItem);
    const renderedFiles = files.map(filename =>
      template(fileItemsTemplate)({ filename })
    );
    const commentTemplate = Template.get(Templates.comment);
    let comment = template(commentTemplate)({
      files: renderedFiles.join("\n")
    });

    const errorCount = errors.length;
    if (errorCount > 0) {
      const errTemplate = Template.get(Templates.errors);
      const errDoc = template(errTemplate)({ errors: errorCount });
      comment = comment + "\n" + errDoc;
    }

    return this.tagComment(comment, CommentTags.Main);
  }

  tagComment(body, commentTag: CommentTags) {
    if (body.indexOf(commentTag) < 0) {
      return body + "\n" + commentTag;
    }
    return body;
  }
}
