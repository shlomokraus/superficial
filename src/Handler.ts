import { Context } from "probot";
import { Parser } from "./Parser";
import fileExtension from "file-extension";
import { Template, Templates } from "./Template";
import { template } from "lodash";
import { Github } from "./Github";
const extensions = ["ts", "js", "tsx", "jsx", "json"];
const botIdentifier = "superficialbot[bot]";
// const botId = 43469792;

export class Handler {
  private readonly context: Context;
  private readonly githubHelper;
  private readonly pr;

  constructor(context: Context, pr) {
    this.context = context;
    this.githubHelper = new Github(context, pr);
    this.pr = pr;
  }

  async handle() {
    const { problematic, errors } = await this.check();

    await this.updateStatus(problematic.length === 0);
    await this.postComment(problematic, errors);

    const revertPaths = problematic.map(item => item.file);
    const revert = await Promise.all(
      revertPaths.map(async path => this.revertFile(path))
    );
    if (revert.length > 0) {
        console.log(this.githubHelper.type);
   // await this.githubHelper.createCommit(revert);
   //   await this.postRevertComment(revert.map(file => file.path));

    }
  }

  async check() {
    const files = await this.getFiles();
    const results = await Promise.all(
      files.map(async file => this.parseFile(file))
    );
    const problematic = results.filter(result => !result.valid);
    const errors = results.filter(result => result.error);
    return { problematic, errors };
  }

  parseFile = async (file: string) => {
    let valid = true;
    let error;

    try {
      const content = await this.getBaseAndHead(file);
      // If we don't have base, that is a new file so it is valid
      valid =
        content.base && content.head
          ? this.compareFiles(content.head, content.base, file)
          : true;
    } catch (ex) {
      this.context.log.error("Unable to parse file", ex);
      error = true;
    }

    return { file, valid, error };
  };

  revertFile = async (path: string) => {
    const base = this.pr.base.ref;
    const source = await this.getFile(path, base);
    return { path, content: source.content };
  };

  postRevertComment = async (files: string[]) => {
    const fileItemsTemplate = await Template.get(Templates.fileItem);
    const renderedFiles = files.map(filename =>
      template(fileItemsTemplate)({ filename })
    );
    const commentTemplate = await Template.get(Templates.revert);
    const comment = template(commentTemplate)({
      files: renderedFiles.join("\n")
    });
    await this.context.github.issues.createComment(
      this.context.repo({ number: this.pr.number, body: comment })
    );
  };

  postComment = async (
    items: { file: string; valid: boolean }[],
    errors: { file: string; error: string }[] = []
  ) => {
    const context = this.context;
    const files = items.map(item => item.file);
    let comment = await this.compileComment(files, errors);

    const existing = await this.getExistingComment();
    if (existing) {
      await context.github.issues.editComment(
        context.repo({
          number: this.pr.number,
          body: `_(repository was updated since posting this message)_\n ${existing.body}`,
          comment_id: String(existing.id)
        })
      );
    } else {
      if (files.length > 0) {
        await context.github.issues.createComment(
          context.repo({ number: this.pr.number, body: comment })
        );
      }
    }
  };

  getExistingComment = async () => {
    const comments = await this.context.github.issues.getComments(
      this.context.repo({number: this.pr.number})
    );
    const filtered = comments.data.find(comment => {
      const isBot = comment.user.login === botIdentifier;
      if (!isBot) {
        return false;
      }

      const body = comment.body;
      return body.indexOf("be superficial!") >= 0;
    });
    return filtered;
  };

  async compileComment(files: string[], errors: any[]) {
    const fileItemsTemplate = await Template.get(Templates.fileItem);
    const renderedFiles = files.map(filename =>
      template(fileItemsTemplate)({ filename })
    );
    const commentTemplate = await Template.get(Templates.comment);
    let comment = template(commentTemplate)({
      files: renderedFiles.join("\n")
    });

    const errorCount = errors.length;
    if (errorCount > 0) {
      const errTemplate = await Template.get(Templates.errors);
      const errDoc = template(errTemplate)({ errors: errorCount });
      comment = comment + "\n" + errDoc;
    }
    return comment;
  }

  compareFiles(source: string, target: string, filename: string) {
    const left = Parser.parse(source, filename);
    const right = Parser.parse(target, filename);
    const diff = Parser.diff(left, right, ["loc", "start", "end"]);
    return diff !== undefined;
  }

  updateStatus = async (success: boolean) => {
    const { head } = this.pr;

    const state = success ? "success" : "failure";

    function getDescription() {
      if (success) return "ready to merge";
      return "superficial changes not allowed";
    }

    let status = {
      sha: head.sha,
      state,
      description: getDescription(),
      context: "Superficial"
    };
    await this.context.github.repos.createStatus(this.context.repo(
      status
    ) as any);
  };

  getFiles = async () => {
    const filesRaw = await this.context.github.pullRequests.getFiles(
      this.context.repo({ number: this.pr.number})
    );
    const files = filesRaw.data.map(file => file.filename);
    const filtered = this.filterFiles(files);
    return filtered;
  };

  getBaseAndHead = async (path: string) => {
    let head;
    let base;
    try {
      const headRef = this.pr.head.ref;
      head = await this.getFileContent(path, headRef);
      const baseRef = this.pr.base.ref;
      base = await this.getFileContent(path, baseRef);
    } catch (ex) {
      ex = ex;
    }

    return { head, base };
  };

  getFileContent = async (path: string, ref: string) => {
    const result = await this.getFile(path, ref);
    const content = Buffer.from(result.content, "base64").toString();
    return content;
  };

  getFile = async (path: string, ref: string) => {
    const repo = this.context.repo({ path, ref });
    const result = await this.context.github.repos.getContent(repo);
    return result.data;
  };

  filterFiles(files: string[]) {
    return files.filter(file => {
      const ext = fileExtension(file);
      return extensions.indexOf(ext) >= 0;
    });
  }
}
