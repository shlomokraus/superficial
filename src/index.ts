import { Handler } from "./Handler";
import { Application } from "probot";

const events = [
  "issue_comment.edited",
  "pull_request.opened",
  "check_suite.requested",
  "check_suite.rerequested"
];

export = async (app: Application) => {

  app.log("Starting superficial bot");
  app.log("Registering events", events);
  
  app.on(events, async context => {
    let prs = getPrs(context);
    await Promise.all(
      prs.map(async pr => {
        const handler = new Handler(context);
        return handler.handle(pr.number);
      })
    );
  });

  const router = app.route('/')

  // Use any middleware
  router.use(require('express').static('static'))

};

function getPrs(context) {
    let prs:any[] = [];
    if (
        context.name === "issue_comment" &&
        context.payload.action === "edited"
      ) {
        prs = [{ number: context.payload.issue.number }];
      } else if (
        context.name === "pull_request" &&
        context.payload.action === "opened"
      ) {
        prs = [context.payload.pull_request];
      } else {
        prs = context.payload.check_suite.pull_requests;
      }
    return prs;
}
