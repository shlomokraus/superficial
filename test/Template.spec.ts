import { Template, Templates} from "../src/Template";

describe("Template", () => {

    it("Should load template file", async () => {
        const str = await Template.get(Templates.comment);
        expect(str).toMatchSnapshot();
    });
})