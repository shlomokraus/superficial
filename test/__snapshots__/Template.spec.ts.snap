// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Template Should load template file 1`] = `
"# Don't be [SUPERFICIAL](https://superficial.herokuapp.com)!

#### This pull request contains files that have formatting changes without any functionality change. 

Those files can make the review process and tracking of commit history much harder. Commits of formatting-only should be submitted in their own PR.  

> You are getting this message because [the superficial bot](https://github.com/shlomokraus/superficial) is installed on this repository.



## Let me help you with that :put_litter_in_its_place: 
Superficial bot can revert to base all files that don't have any change to their functionality. 
Click this checkbox to trigger an automatic commit: 

- [ ] Remove selected files (this may take a few minutes)

#### The following files should be removed:

Path  | 
---------| 
<%= files %>

---
"
`;
